### ISSUE TITLE
Add the ability to rotate items and their children

### ISSUE THREAD

**Issue by tronical** [slint-ui/slint#1481]
Update: https://github.com/slint-ui/slint/issues/1481#issuecomment-3223204716

---

For certain types of UIs - like for example a compass - we need the ability to rotate an item and its children. https://github.com/slint-ui/slint/pull/1478 revives some old code to bring back this feature as:

    `rotation-angle`
    `rotation-origin-x`
    `rotation-origin-y`

properties on any items, resulting in an injected hidden `Rotate` element that applies the rotation when rendering.

Commit 7b50e38e533289c1fd05150715956a7930902397 implements the rotation also for the Skia backend.

There are a few things left before this issue can be considered complete:

- [ ] The API needs to be reviewed / discussed. Is this the correct approach, or do we need a more general "transform" API? Could we start with this and generalise later?
- [ ] Input events need to be subjected to the rotation as well.
- [ ] These properties need to be documented, preferably with an inline example.

**Comment by ogoffart** [comment]
Regarding the origin, i wonder if it would not be better to have, instead of

```
rotation-origin-x: width / 2;
rotation-origin-y: height / 2;
```

Maybe something like one property of type point:

``` 
rotation-origin: { x: width / 2, y: height / 2 };
```

One day if we add tuple to the langage, we could even shorten it to  `rotation-origin: (width / 2, height / 2);`
But what we can already do is use the % as `rotation-origin: { x: 50%, y: 50% };` although that would then be easier with two different properties.

**Comment by ogoffart** [comment]
Maybe it is also better in the first release to limit rotation to `Image` elements, that do not have any children.

This would solve the issue with translating input events.
And this would make it easier for the implementation in the software renderer if only images can be rotated.
(Rotating Rectangle (possibily with border), Clip and Text is each some work to do) 

Image rotation is the primary use case for this feature anyway.

**Comment by tronical** [comment]
> Regarding the origin, i wonder if it would not be better to have, instead of
> 
> ```
> rotation-origin-x: width / 2;
> rotation-origin-y: height / 2;
> ```
> 
> Maybe something like one property of type point:
> 
> ```
> rotation-origin: { x: width / 2, y: height / 2 };
> ```

Yes. I think this would apply to quite a few places where we have `foo-bar-baz` that would be better written as

`foo-bar: { baz: 42; blub: 100; }`

I've been thinking that this is something that we could perhaps implement generically in the compiler. We maintain to have individual `rotation-origin-x` and `rotation-origin-y` properties in the implementation, but we allow for the syntactic sugar to write `rotation-origin: { x: ...; y: ... }`.

If for some reason that's a little too "magic" then we could at least annotate these kind of property groups where we allow that in builtins.slint, for example for all the `font-XXX` properties.

> One day if we add tuple to the langage, we could even shorten it to `rotation-origin: (width / 2, height / 2);` But what we can already do is use the % as `rotation-origin: { x: 50%, y: 50% };` although that would then be easier with two different properties.
> 
> Edit: Should the default not be (50%,50%) anyway?

Yes!!

**Comment by tronical** [comment]
> Maybe it is also better in the first release to limit rotation to `Image` elements, that do not have any children.
> 
> This would solve the issue with translating input events. And this would make it easier for the implementation in the software renderer if only images can be rotated. (Rotating Rectangle (possibily with border), Clip and Text is each some work to do)
> 
> Image rotation is the primary use case for this feature anyway.

For the record, that's fine with me :)

**Comment by ogoffart** [comment]
Now it is implemented and documented but it only work on childless Images.
I keep this open for rotating arbitrary item that may include clip and mouse event and the like.

**Comment by flukejones** [comment]
I would like to see this for text. But also for arbitrary widgets since it will be a requirement on Android and iOS.

**Comment by flukejones** [comment]
Actually a specific use I have right now for text is to rotate to label a Y axis in a graph.

**Comment by lmnewey** [comment]
I would like to see the progress bar rotatable, as the above has stated labels on other axis is another use case

**Comment by maciek134** [comment]
So if I understand this correctly, the main things preventing rotation from working outside images are:
- software renderer
- event handling

right?

Disabling the check and modifying examples a bit, rendering seems fine on Linux:
![image](https://github.com/slint-ui/slint/assets/1146309/cc3bb2db-9c22-4507-b16c-68c594b36220)
but ofc the button barely works.

I'd be willing to make a PR with the necessary changes if nobody is working on that.

**Comment by tronical** [comment]
The ability to rotate Text element that don't have children is implemented now for FemtVG/Skia/Qt (as part of #5532).

**Comment by maciek134** [comment]
Thanks @ogoffart , makes sense.

In that case I'll start with arbitrary element rotation with input and follow up with the software renderer implementation later - the math for dirty region calculation should have a lot of overlap with event handling.

**Comment by aurexav** [comment]
Would like to have this.

For example, I can use two border radius rectangle `/` `\` to make a nice `<`.

**Comment by ASxa86** [comment]
Are there any recommended workarounds for this in the mean time.

I have an App in QML where I define a rectangle with a left and right button with text in the center. I would like to rotate this rectangle vertically so the buttons are vertical along with the text. I then take this rectangle and make many more that I layout in a grid for table top life tracking in board/card games.

In QML, this is relatively easy to accomplish as I can parent all these to a single item I rotate.

The only thing I could think to do right now is create unique implementations for each rotation I plan to support which is inconvenient.

**Comment by jwtocadero** [comment]
As another motivation, as this would directly also enable scaling, it is quiet common to simulate haptic feedback in UIs by scaling it when pressed. I think this feature is very urgent to implement.

**Comment by ogoffart** [comment]
As of now, `Text` and `Image` without children are the only elements that can be rotated.
As explained before the reason we can't rotate arbitrary element is mostly because the mapping of input hasn't been implemented.

As mentioned in a [previous comment](https://github.com/slint-ui/slint/issues/1481#issuecomment-2184132969), if we remove the check from the compiler, it should render correctly with all our renderer, (except with the software renderer)
https://github.com/slint-ui/slint/blob/d6846c352a2dc618d3155a3b530eeef0db6bdf89/internal/compiler/passes/check_rotation.rs#L9-L10

The partial renderer that is used by Skia in some case would need to be checked that the item bounds are correct if there is rotation. (`PartialRenderer` in item_rendering.rs)

In order to add input mapping, we'd need to add something in `input.rs` in the `send_mouse_event_to_item` we do apply a translation, but we would somehow need to apply rotation as well. The simpler method would probably to use `downcast::<Rotate>` to figure out the rotation. Other option may include adding an extra transfrom or rotation function like we have a `geometry` function, in one of the trait (`Item` or `ItemTree`)

I think it would be great to have support for rotated and maybe smooth scaling of image in the software renderer. If we have arbitrary item that means we'll also need to have rotation of Rectangle (including rounded ones)

[This comment](https://github.com/slint-ui/slint/issues/1481#issuecomment-2186078396) is still valid  (Minus the Text part which was implemented already)

**Comment by redstrate** [comment]
Now that #9317 is merged, it looks like the only thing that's left is implementing image rotation support in software rendering correct?

**Comment by ogoffart** [comment]
Right, this is done in 1.14

The missing implementation for the software renderer is tracker in https://github.com/slint-ui/slint/issues/3068
